File: MYCM/support/persistence/repositories/DataRepository.cs
using System.Collections.Generic;

namespace support.persistence.repositories{
    /// <summary>
    /// Generic Interface for CRUD functionalities on a data repository
    /// </summary>
    /// <typeparam name="E">Generic-Type of the repository entity</typeparam>
    /// <typeparam name="ID">Generic-Type of the repository entity persistence identity</typeparam>
    public interface DataRepository<E,ID>{
        /// <summary>
        /// Saves an entity on the repository
        /// </summary>
        /// <param name="entity">E with the entity being saved on the repository</param>
        /// <returns>E with the saved entity</returns>
        E save(E entity);

        /// <summary>
        /// Updates the state of an entity on the repository
        /// </summary>
        /// <param name="entity">E with the entity being updated on the repository</param>
        /// <returns>E with the updated entity</returns>
        E update(E entity);
        
        /// <summary>
        /// Removes an entity from the repository
        /// </summary>
        /// <param name="entity">E with the entity being removed from the repository</param>
        /// <returns>E with the removed entity</returns>
        E remove(E entity);
        
        /// <summary>
        /// Finds an entity based on persistence identifier
        /// </summary>
        /// <param name="entityPersistenceID">ID with the entity persistence identifier</param>
        /// <returns>E with the entity which is represented by a certain persistence identifier</returns>
        E find(ID entityPersistenceID);
        /// <summary>
        /// Finds all entities of a repository
        /// </summary>
        /// <returns>IEnumerable with all of the entities of a repository</returns>
        IEnumerable<E> findAll();
    }
}

File: MYCA/myca/support/src/main/java/cdiomyc/support/persistence/repositories/DataRepository.java
package cdiomyc.support.persistence.repositories;

/**
 * Generic Interface that represents all simple functionalities regarding data 
 * actions in repositories
 * @author <a href="https://github.com/freitzzz">freitzzz</a>
 * @param <E> Generic-Type of the repository entity
 * @param <ID> Generic-Type of the repository entity persistence identifier
 */
public interface DataRepository<E,ID> {
    /**
     * Method that add's a certain entity to the current entity repository
     * @param entity T with the Generic entity being added to the repository
     * @return T with the added entity
     */
    public abstract E save(E entity);
    
    /**
     * Method that updates a certain entity from the current entity repository
     * @param entity T with the Generic entity being updated of the repository
     * @return T with the updated entity
     */
    public abstract E update(E entity);
    /**
     * Method that removes a certain entity from the current entity repository
     * @param entity T with Generic entity being removed from the repository
     * @return T with the removed entity
     */
    public abstract E remove(E entity);
    /**
     * Method that finds a certain entity from the current entity repository 
     * using it's persistence identifier
     * @param id ID with the Generic entity persistence identifier
     * @return T with the entity which is represented by a certain persistence 
     * identifier
     */
    public abstract E find(ID id);
    /**
     * Method that finds all entities from the current entity repository
     * @return Iterable with all the entities from the current repository
     */
    public abstract Iterable<E> findAll();
}
